src/Interpreters/loadMetadata.cpp

```diff
@@ -495,6 +501,116 @@ void convertDatabasesEnginesIfNeed(const LoadTaskPtrs & load_metadata, ContextMu
+    const auto & as_create = as_create_ptr->as<ASTCreateQuery &>();
+    const auto & storage = as_create.storage->as<ASTStorage &>();
+    String storage_definition = queryToString(storage);
+
+    /// Get storage definition
+    /// Set uuid explicitly, because it is forbidden to use the 'uuid' macro without ON CLUSTER
+    auto uuid = UUIDHelpers::generateV4();
+    String replica_path = context->getConfigRef().getString("default_replica_path", "/clickhouse/tables/{uuid}/{shard}");
+    replica_path = boost::algorithm::replace_all_copy(replica_path, "{uuid}", fmt::format("{}", uuid));
+    String replica_name = context->getConfigRef().getString("default_replica_name", "{replica}");
+    String replicated_args = fmt::format("('{}', '{}')", replica_path, replica_name);
+    String replicated_engine = "Replicated" + storage.engine->name + replicated_args;
+
+    storage_definition = boost::algorithm::replace_first_copy(storage_definition, storage.engine->name, replicated_engine);
+
+    /// Get names
+    String table_name = id.getTableName();
+    String database_name = id.getDatabaseName();
+    String qualified_quoted_name = id.getFullTableName();
+    id.table_name = id.table_name + "_temp";
+    String tmp_qualified_quoted_name = id.getFullTableName();
+
+    try
+    {    
+        String create_table_query = fmt::format("CREATE TABLE {} UUID '{}' AS {} {}", tmp_qualified_quoted_name, uuid, qualified_quoted_name, storage_definition);
+        auto res = executeQuery(create_table_query, context, QueryFlags{ .internal = true }).second;
+        executeTrivialBlockIO(res, context);
+
+        String exchange_tables_query = fmt::format("EXCHANGE TABLES {} AND {}", qualified_quoted_name, tmp_qualified_quoted_name);
+        res = executeQuery(exchange_tables_query, context, QueryFlags{ .internal = true }).second;
+        executeTrivialBlockIO(res, context);
+
+        /// Get partition ids
+        String get_attach_queries_query = fmt::format("SELECT DISTINCT partition_id FROM system.parts WHERE table = '{}' AND database = '{}' AND active;", id.table_name, database_name);
+        WriteBufferFromOwnString buffer2;
+        ReadBufferFromOwnString buffer3 {std::move(get_attach_queries_query)};
+        auto select_query_context = Context::createCopy(context);
+        select_query_context->makeQueryContext();
+        select_query_context->setCurrentQueryId("");
+
+        executeQuery(buffer3, buffer2, false, select_query_context, {}, {.internal=true});
+
+        std::stringstream partition_ids_string{buffer2.str()};
+        std::string line;
+
+        /// Attach partitions
+        while (std::getline(partition_ids_string, line, '\n'))
+        {
+            String query3 = fmt::format("ALTER TABLE {} ATTACH PARTITION ID '{}' FROM {};", qualified_quoted_name, line, tmp_qualified_quoted_name);
+            executeQuery(query3, select_query_context, {.internal=true});
+        }
```

Comment:
FYI there's `SYSTEM RESTORE REPLICA` query and `StorageReplicatedMergeTree::restoreMetadataInZooKeeper()`, please consider using it (just update the engine and restore)

Reply:
I think this won't help attach partitions from the old table since there is nothing inside new table's directory even after name exchange

Reply:
I mean that you don't need a new table and you don't need to exchange them. You can just update the engine of the old table, load it as ReplicatedMergeTree (it will be in readonly mode due to missing metadata in zk) and restore it.